remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java

```diff
@@ -0,0 +1,171 @@
+                    logSegmentData.leaderEpochIndex().array());
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.ProducerSnapshot),
+                    Files.readAllBytes(logSegmentData.producerSnapshotIndex().toPath()));
+        } catch (IOException e) {
+            throw new RemoteStorageException(e.getMessage(), e);
+        }
+        log.debug("copied log segment and indexes for : {} successfully.", remoteLogSegmentMetadata);
+    }
+
+    @Override
+    public InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                       int startPosition)
+            throws RemoteStorageException {
+        log.debug("Received fetch segment request at start position: [{}] for [{}]", startPosition, remoteLogSegmentMetadata);
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata can not be null");
+
+        return fetchLogSegment(remoteLogSegmentMetadata, startPosition, Integer.MAX_VALUE);
+    }
+
+    @Override
+    public InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                       int startPosition,
+                                       int endPosition) throws RemoteStorageException {
+        log.debug("Received fetch segment request at start position: [{}] and end position: [{}] for segment [{}]",
+                startPosition, endPosition, remoteLogSegmentMetadata);
+
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata can not be null");
+
+        if (startPosition < 0 || endPosition < 0) {
+            throw new IllegalArgumentException("Given start position or end position must not be negative.");
+        }
+
+        if (endPosition < startPosition) {
+            throw new IllegalArgumentException("end position must be greater than start position");
+        }
+
+        String key = generateKeyForSegment(remoteLogSegmentMetadata);
+        byte[] segment = keyToLogData.get(key);
+
+        if (segment == null) {
+            throw new RemoteResourceNotFoundException("No remote log segment found with start offset:"
+                                                      + remoteLogSegmentMetadata.startOffset() + " and id: "
+                                                      + remoteLogSegmentMetadata.remoteLogSegmentId());
+        }
+
+        if (startPosition >= segment.length) {
+            throw new IllegalArgumentException("start position: " + startPosition
+                                               + " must be less than the length of the segment: " + segment.length);
+        }
+
+        // check for boundaries like given end position is more than the length, length should never be more than the
+        // existing segment size.
+        int length = Math.min(segment.length - 1, endPosition) - startPosition + 1;
```

Comment:
Hmm, do we need to explicitly check if `endPosition` < `segment.length`?

Reply:
It takes `Math.min(endPosition, segment.length)`. So, no need to have that check.